datafusion/physical-expr/src/expressions/case.rs

```diff
@@ -964,24 +1064,38 @@ impl CaseBody {
         &self,
         batch: &RecordBatch,
         when_value: &BooleanArray,
-        return_type: &DataType,
     ) -> Result<ColumnarValue> {
-        let then_value = self.when_then_expr[0]
-            .1
-            .evaluate_selection(batch, when_value)?
-            .into_array(batch.num_rows())?;
+        let when_value = match when_value.null_count() {
+            0 => Cow::Borrowed(when_value),
+            _ => {
+                // `prep_null_mask_filter` is required to ensure null is treated as false
+                Cow::Owned(prep_null_mask_filter(when_value))
+            }
+        };
+
+        let optimize_filter = batch.num_columns() > 1;
```

Comment:
It might also be worth checking if there are any nested types (e.g. structarrays) and optimize the filter in that case too -- this is done elsewhere (maybe in the filter kernel itself 🤔 )

Reply:
Agreed. The logic that handles that isn't `pub` in arrow-rs unfortunately. I can duplicate it here if you like.

Reply:
Fixed for now. Would it be useful to make this a method of DataType? I can prepare an `arrow-rs` PR for that.